src/renate/memory/buffer.py

```diff
@@ -104,53 +33,47 @@ class DataBuffer(Dataset, ABC):
     data point in the buffer. Additional fields of metadata might be added by some buffering
     methods, e.g., instance weights in coreset methods.
 
+    In addition to passing metadata to the `update` method, one also access and replace the metadata
+    in the buffer via the `get_metadata` and `set_metadata` methods.
+
     Note that the buffer does not change the device placement of data passed to it. Please ensure
     that the data passed to `DataBuffer.update` resides on the CPU.
 
     Note that, in order to apply transformations, the buffer assumes that the data points are tuples
     of the form `(x, y)`. We apply `transform` to `inputs` and `target_transform` to `y`. Ensure
     that the transforms accept the correct type, e.g., if `x` is a dictionary, `transform` needs to
     operate on a dictionary.
-
-    Args:
-        max_size: Maximal size of the buffer.
-        storage_mode: How to store the data in the buffer. Currently, we only support `in_memory`.
-        seed: Seed for the random number generator used in the buffer.
-        transform: The transformation to be applied to the inputs "x" of points in the buffer.
-        target_transform: The transformation to be applied to target "y" of points in the buffer.
     """
 
     def __init__(
         self,
         max_size: Optional[int] = None,
-        storage_mode: defaults.SUPPORTED_BUFFER_STORAGE_MODES = defaults.BUFFER_STORAGE_MODE,
         seed: int = defaults.SEED,
         transform: Optional[Callable] = None,
         target_transform: Optional[Callable] = None,
     ) -> None:
         self._max_size = max_size
-        self._storage_mode = storage_mode
         self._seed = seed
+        self._transform = transform
+        self._target_transform = target_transform
 
         self._rng = get_generator(self._seed)
 
         self._count = 0
-
-        self._data_points = None
-        self.metadata = None
-        self._size = 0
-
-        self._transform = transform
-        self._target_transform = target_transform
+        self._datasets = []
+        self._indices = {}
+        self._data_point_prototype = None
```

Comment:
We  keep around a prototypical data point to infer shape and data type. That is because the memory-mapped tensors that we use for storage don't save that information. The prototype just becomes part of the buffer's state dict, so we can re-infer this info.